### PR TITLE
Fix listing of modified files

### PR DESCRIPTION
**Pull request by vkucera**
Consider the merge-base commit as the starting point of diff instead of listing all differences between branches.

Fixes https://github.com/oxsecurity/megalinter/issues/2125

## Proposed Changes

MegaLinter currently gets the list of files to check by comparing `HEAD` and the base branch.

- That is fine for `pull_request` events where `HEAD` is the merge commit and `git diff` only lists files in the PR.
- It is not fine for `push` events where the base branch can contain commits not present in the head branch and `git diff` also lists files from those commits.

To list only files from commits in the head branch which are not present in the base branch, one should run `git diff` against the common ancestor (aka merge-base) of the two branches.
This can be done with `git diff <base-branch>...` which is equivalent to:

- `git diff <base-branch>...HEAD`
- `git diff --merge-base <base-branch> HEAD`
- `git diff $(git merge-base <base-branch> HEAD) HEAD`

### Author/Contributor
- [x] Add entry to the [CHANGELOG](https://github.com/oxsecurity/megalinter/blob/main/CHANGELOG.md) listing the change and linking to the corresponding issue (if appropriate)
- [ ] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

**Comment by vkucera** [comment]
The test failed after [5e1adac](https://github.com/oxsecurity/megalinter/pull/3472/commits/5e1adaca3495dbb284da718e8ac6e0b2ccc38c10) because of a missing merge base which I fixed by adding `fetch-depth: 0` in the checkout step of that action.
Now the test is failing at `test_get_linter_help` but I don't understand the issue.

**Comment by nvuillam** [comment]
it seems everything is working :)

We'll merge this one once the current CI issue with eslint-plugin-json is fixed :)

**Comment by vkucera** [comment]
> CI issue

Thanks @nvuillam , which issue are you referring to?

**Comment by vkucera** [comment]
> it seems everything is working :)
> 
> We'll merge this one once the current CI issue with eslint-plugin-json is fixed :)

Thanks @nvuillam , which issue are you referring to?

**Comment by nvuillam** [comment]
@vkucera it's fixed, i merge once CI is ok (and it should be ^^ )

File: .github/workflows/deploy-DEV.yml [hunk 1]
```diff
@@ -63,6 +63,8 @@ jobs:
     steps:
       - name: Checkout Code
         uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (echoix) [inline .github/workflows/deploy-DEV.yml:67]: @vkucera You know that here is only for our CI, not the templates that are generated by the mega-linter-runner when a user first installs Megalinter. Is that what you wanted?

Comment (vkucera) [reply .github/workflows/deploy-DEV.yml:67]: @echoix I added it to pass the CI. Please see https://github.com/oxsecurity/megalinter/pull/3472#issuecomment-2042200621

Comment (echoix) [reply .github/workflows/deploy-DEV.yml:67]: Ok good!

**Review by echoix — APPROVED** [review]

**Pull request merged**